File: testdata/jennies/builders/struct_with_defaults/PythonBuilder/builders/struct_with_defaults.py
import typing
from ..cog import builder as cogbuilder
from ..models import struct_with_defaults


class NestedStruct(cogbuilder.Builder[struct_with_defaults.NestedStruct]):
    _internal: struct_with_defaults.NestedStruct

    def __init__(self):
        self._internal = struct_with_defaults.NestedStruct()

    def build(self) -> struct_with_defaults.NestedStruct:
        """
        Builds the object.
        """
        return self._internal    
    
    def string_val(self, string_val: str) -> typing.Self:    
        self._internal.string_val = string_val
    
        return self
    
    def int_val(self, int_val: int) -> typing.Self:    
        self._internal.int_val = int_val
    
        return self
    


class Struct(cogbuilder.Builder[struct_with_defaults.Struct]):
    _internal: struct_with_defaults.Struct

    def __init__(self):
        self._internal = struct_with_defaults.Struct()

    def build(self) -> struct_with_defaults.Struct:
        """
        Builds the object.
        """
        return self._internal    
    
    def all_fields(self, all_fields: cogbuilder.Builder[struct_with_defaults.NestedStruct]) -> typing.Self:    
        all_fields_resource = all_fields.build()
        self._internal.all_fields = all_fields_resource
    
        return self
    
    def partial_fields(self, partial_fields: cogbuilder.Builder[struct_with_defaults.NestedStruct]) -> typing.Self:    
        partial_fields_resource = partial_fields.build()
        self._internal.partial_fields = partial_fields_resource
    
        return self
    
    def empty_fields(self, empty_fields: cogbuilder.Builder[struct_with_defaults.NestedStruct]) -> typing.Self:    
        empty_fields_resource = empty_fields.build()
        self._internal.empty_fields = empty_fields_resource
    
        return self
    
    def complex_field(self, complex_field: unknown) -> typing.Self:    
        self._internal.complex_field = complex_field
    
        return self
    
    def partial_complex_field(self, partial_complex_field: unknown) -> typing.Self:    
        self._internal.partial_complex_field = partial_complex_field
    
        return self
    

File: testdata/jennies/builders/struct_with_defaults/RustBuilder/src/builders/struct_with_defaults.rs
use crate::cog;
use crate::types::struct_with_defaults::NestedStruct;
use crate::types::struct_with_defaults::Struct;

#[derive(Debug, Clone)]
pub struct NestedStructBuilder {
    internal: NestedStruct,
    errors: Vec<cog::BuildError>,
}

impl NestedStructBuilder {
    pub fn new() -> Self {
        Self {
            internal: NestedStruct::default(),
            errors: Vec::new(),
        }
    }
}

impl Default for NestedStructBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl NestedStructBuilder {
    pub fn string_val(mut self, string_val: String) -> Self {
        self.internal.string_val = string_val;

        self
    }
}

impl NestedStructBuilder {
    pub fn int_val(mut self, int_val: i64) -> Self {
        self.internal.int_val = int_val;

        self
    }
}

impl cog::Builder<NestedStruct> for NestedStructBuilder {
    fn build(&self) -> Result<NestedStruct, Vec<cog::BuildError>> {
        if !self.errors.is_empty() {
            return Err(self.errors.clone());
        }

        Ok(self.internal.clone())
    }
}

#[derive(Debug, Clone)]
pub struct StructBuilder {
    internal: Struct,
    errors: Vec<cog::BuildError>,
}

impl StructBuilder {
    pub fn new() -> Self {
        Self {
            internal: Struct::default(),
            errors: Vec::new(),
        }
    }
}

impl Default for StructBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl StructBuilder {
    pub fn all_fields(mut self, all_fields: impl cog::Builder<NestedStruct>) -> Self {
        let built0 = match all_fields.build() {
            Ok(val) => val,
            Err(mut err) => {
                self.errors.append(&mut err);
                return self;
            }
        };
        self.internal.all_fields = built0;

        self
    }
}

impl StructBuilder {
    pub fn partial_fields(mut self, partial_fields: impl cog::Builder<NestedStruct>) -> Self {
        let built0 = match partial_fields.build() {
            Ok(val) => val,
            Err(mut err) => {
                self.errors.append(&mut err);
                return self;
            }
        };
        self.internal.partial_fields = built0;

        self
    }
}

impl StructBuilder {
    pub fn empty_fields(mut self, empty_fields: impl cog::Builder<NestedStruct>) -> Self {
        let built0 = match empty_fields.build() {
            Ok(val) => val,
            Err(mut err) => {
                self.errors.append(&mut err);
                return self;
            }
        };
        self.internal.empty_fields = built0;

        self
    }
}

impl StructBuilder {
    pub fn complex_field(mut self, complex_field: serde_json::Value) -> Self {
        self.internal.complex_field = complex_field;

        self
    }
}

impl StructBuilder {
    pub fn partial_complex_field(mut self, partial_complex_field: serde_json::Value) -> Self {
        self.internal.partial_complex_field = partial_complex_field;

        self
    }
}

impl cog::Builder<Struct> for StructBuilder {
    fn build(&self) -> Result<Struct, Vec<cog::BuildError>> {
        if !self.errors.is_empty() {
            return Err(self.errors.clone());
        }

        Ok(self.internal.clone())
    }
}
